### PR TITLE
Improve error message after failed RMM shutdown

### PR DESCRIPTION
**Pull request by andygrove**
Exceptions during `GpuDeviceManager.shutdown` can lead to `GpuDeviceManager` and ` RapidsBufferCatalog` being left in an inconsistent state which results in `NullPointerException` when accessing uninitialized state.

This PR changes the `singletonMemoryInitialized` flag from a boolean to a `MemoryState` data type with possible values of `Initialized`, `Uninitialized`, and `Uninitializable`, with appropriate error checking added.

Signed-off-by: Andy Grove <andygrove@nvidia.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 1]
```diff
@@ -28,6 +28,12 @@ import org.apache.spark.internal.Logging
 import org.apache.spark.resource.ResourceInformation
 import org.apache.spark.sql.rapids.GpuShuffleEnv
 
+sealed trait MemoryState
+case object Initialized extends MemoryState
+case object Uninitialized extends MemoryState
+case object Uninitializable extends MemoryState
+
+
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:36]: extra space nit.

**Review by abellina — APPROVED** [review]
LGTM @andygrove, tiny nit.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala [hunk 2]
```diff
@@ -127,9 +133,17 @@ object GpuDeviceManager extends Logging {
...
+      RapidsBufferCatalog.close()
+      Rmm.shutdown()
+      singletonMemoryInitialized = Uninitialized
+    } catch {
+      case e: Throwable =>
+        // flag as Uninitializable so that any future initialization attempts can throw
+        // a helpful error message
+        singletonMemoryInitialized = Uninitializable
+        throw e
+    }
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala:145]: Can simplify this by assuming shutdown will fail up front
```suggestion
    // assume error during shutdown until we complete it
    singletonMemoryInitialied = Uninitializable
    RapidsBufferCatalog.close()
    Rmm.shutdown()
    singletonMemoryInitialized = Uninitialized
```

**Review by jlowe — APPROVED** [review]

**Review by abellina — APPROVED** [review]

**Comment by abellina** [comment]
build

**Pull request merged**